### PR TITLE
Split warnings and info annotations in API response

### PR DESCRIPTION
**Pull request by zenador**
Addresses https://github.com/prometheus/prometheus/issues/14135

### CODE REVIEW

**Review by bboreham — DISMISSED** [review]
This change lgtm.

The existing implementation by which `errors.Is(err, PromQLInfo)` is working seems a bit tricksy to me; better done via Go types.

**Comment by beorn7** [comment]
> The existing implementation by which `errors.Is(err, PromQLInfo)` is working seems a bit tricksy to me; better done via Go types.

I guess this was my idea, seduced by the convenience of using `fmt.Errorf` with `%w`. It could be changed to a different approach, but obviously not as part of this PR.

File: util/annotations/annotations.go [hunk 1]
```diff
@@ -71,27 +71,44 @@ func (a Annotations) AsErrors() []error {
...
+		}
 	}
-	if maxAnnos > 0 && len(a) > maxAnnos {
-		arr = append(arr, fmt.Sprintf("%d more annotations omitted", len(a)-maxAnnos))
+	if warnSkipped > 0 {
+		warnArr = append(warnArr, fmt.Sprintf("%d more warnings omitted", warnSkipped))
 	}
-	return arr
+	if infoSkipped > 0 {
+		infoArr = append(infoArr, fmt.Sprintf("%d more info messages omitted", infoSkipped))
```
Comment (beorn7) [inline util/annotations/annotations.go:109]: messages → annotations

But we should stay consistent in any case, so either change the message in L120 to "%d more warning annotations omitted", or keep L120 as is and change this into "%d more infos omitted". (But I get how the latter reads a bit weird. I would still be fine with it, as it is shorter. Your call.)

File: web/api/v1/api.go [hunk 2]
```diff
@@ -159,6 +159,7 @@ type Response struct {
 	ErrorType errorType   `json:"errorType,omitempty"`
 	Error     string      `json:"error,omitempty"`
 	Warnings  []string    `json:"warnings,omitempty"`
+	Info      []string    `json:"info,omitempty"`
```
Comment (beorn7) [inline web/api/v1/api.go:162]: Maybe `Infos` for consistency?

File: web/api/v1/api.go [hunk 3]
```diff
@@ -1746,11 +1747,13 @@ func (api *API) cleanTombstones(*http.Request) apiFuncResult {
...
 func (api *API) respond(w http.ResponseWriter, req *http.Request, data interface{}, warnings annotations.Annotations, query string) {
 	statusMessage := statusSuccess
+	warn, info := warnings.AsStrings(query, 10, 10)
 
 	resp := &Response{
 		Status:   statusMessage,
 		Data:     data,
-		Warnings: warnings.AsStrings(query, 10),
+		Warnings: warn,
+		Info:     info,
```
Comment (beorn7) [inline web/api/v1/api.go:1756]: I would definitely prefer `Infos` here. The context is clear, i.e. that this is a plural to mark that the value for this key is a list of strings.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thank you very much. Looks good in general. I just have a number of ideas about naming. See individual comments.

**Comment by beorn7** [comment]
The [comment in lines 116 to 119](https://github.com/prometheus/prometheus/pull/14327/files#diff-d8edc673795ebafcce64036e85a031d0fd3d09cbcaea9bb1cc140373542afc92R116-R119) needs an update. (I cannot comment on unchanged lines directly, so I'm providing the link here.)

**Comment by zenador** [comment]
@beorn7 PTAL

**Review by beorn7 — APPROVED** [review]
Looks good. Thank you.

Do you think you could add a "minimal UI change" to this PR? Like rendering the infos in the same way as the warnings, just in a slightly less alarming color. That would be an MVP we could merge, and then we could ask the UI experts to make it nicer later, like including a button to hide infos or something.

**Comment by zenador** [comment]
@beorn7 Added the minimal UI change

File: web/ui/react-app/src/pages/graph/Panel.tsx [hunk 4]
```diff
@@ -307,6 +310,11 @@ class Panel extends Component<PanelProps, PanelState> {
             <Col>{warning && <Alert color="warning">{warning}</Alert>}</Col>
           </Row>
         ))}
+        {this.state.infos?.map((info, index) => (
+          <Row key={index}>
+            <Col>{info && <Alert color="success">{info}</Alert>}</Col>
```
Comment (juliusv) [inline web/ui/react-app/src/pages/graph/Panel.tsx:315]: `success` is green, I'd pick `info` as the info-level color (blue/turquoise-ish):

```suggestion
            <Col>{info && <Alert color="info">{info}</Alert>}</Col>
```

**Comment by juliusv** [comment]
UI change looks good to me besides the comment on the color.

**Review by juliusv — APPROVED** [review]
Thanks 👍

**Pull request merged**